src/vmm/src/builder.rs

```diff
@@ -184,23 +502,19 @@ fn create_vmm_and_vcpus(
     // For x86_64 we need to create the interrupt controller before calling `KVM_CREATE_VCPUS`
     // while on aarch64 we need to do it the other way around.
     #[cfg(target_arch = "x86_64")]
-    let (vcpus, pio_device_manager) = {
-        setup_interrupt_controller(&mut vm)?;
-        let vcpus = create_vcpus(&vm, vcpu_count, &vcpus_exit_evt).map_err(Internal)?;
-
-        // Make stdout non blocking.
-        set_stdout_nonblocking();
-
+    let pio_device_manager = {
         // Serial device setup.
         let serial_device =
-            setup_serial_device(event_manager, std::io::stdin(), io::stdout()).map_err(Internal)?;
+            setup_serial_device(std::io::stdin(), io::stdout()).map_err(Internal)?;
 
         // x86_64 uses the i8042 reset event as the Vmm exit event.
         let reset_evt = vcpus_exit_evt
             .try_clone()
             .map_err(VmmError::EventFd)
             .map_err(Internal)?;
 
+        x86_64::setup_interrupt_controller(&mut vm).map_err(Internal)?;
```

Comment:
Here we can also move creation of interrupt controller into the `create_vmm_and_vcpus`. But with it we need to move the `pio_dev_mgr.register_devices(vm.fd()).unwrap();` line there as well.